sr_port/alias_funcs.c

```diff
@@ -290,10 +293,9 @@ void als_lsymtab_repair(hash_table_mname *table, ht_ent_mname *table_base_orig,
 			}
 		}
 		fpprev = fp;
-		fp = fp->old_frame_pointer;
```

Comment:
This all has to do with the elimination of the GTM_CI frame. The frame chain used to look like this:
CI-base-frame <- GTM$CI <- M routine

Now it looks like:
CI-base-frame <- M routine

In the previous scheme the GTM$CI frame had a "flag value" SFF_CI that indicated this was a call-in frame. In the new scheme the the base frame has a "type value" with SFT_CI set that indicates this is a call-in frame. The removed statement was because once we have noted the frame with the type set for a call-in frame, we no longer have to go back one more frame to get to its base frame. I will a version of this note to the notes for alias_funcs.c.

Reply:
I still don't understand the connection between SFF_CI/SFT_CI and the break that happens in line 297 in the new code. Let us say there is NO call-in frame at all (i.e. no call-ins done in this process). If the break happens because "done" is TRUE, we would come out of the loop with a value of "fp" that is one frame newer than older code. And it is not clear to me if that is correct.

Reply:
```
There are two cases - a call-in frame is present or a call-in frame is NOT present. First the call-in frame is
present case. Skip to "***" to see not present case.

Previous code:

1        do
2        {
3               ... <irrelevant to this discussion stuff>
4               fpprev = fp;
5               fp = fp->old_frame_pointer;
6               if (done)
7                       break;
8               if (SFF_CI & fpprev->flags)
9               {       /* Callins needs to be able to crawl past apparent end of stack to earlier stack segments.
10                       * We should be in the base frame now. See if an earlier frame exists.
11                       * Note we don't worry about trigger base frames here because triggers *always* have a
12                       * different symbol table - previous symbol tables and stack levels are not affected.
13                       */
14                      fp = *(stack_frame **)(fp + 1); /* Backups up to "prev pointer" created by base_frame() */
15                      if ((NULL == fp) || (fp >= (stack_frame *)stackbase) || (fp < (stack_frame *)stacktop))
16                              break;  /* Pointer not within the stack -- must be earliest occurence */
17              }
18      } while(fp);
19      /* Next, check the mv_stents for the stackframes we processed. Certain mv_stents also have hash
20       * table references in them that need repair.
21       */
22      for (mv_st_ent = mv_chain;
23           mv_st_ent < (mv_stent *)(fp ? fp : fpprev);        /* Last stack frame actually processed */
24           mv_st_ent = (mv_stent *)(mv_st_ent->mv_st_next + (char *)mv_st_ent))
25      {
26              ... <rest of loop>
27      }

Recall that the M stack looks like:
[1:M-base-frame (no prev frame)] : [2:callin-base-frame] : [3:GTM$CI-frame (has SFF_CI set)] : [4:callin-routine-frame]

Operations in loop:
  1. fp -> frame 4 (line 1)
  2. fpprev -> frame 4 (line 4)
  3. fp -> frame 3 (line 5)
  4. test on line 8 is FALSE
  5. line 18 loop back up
  6. fpprev = frame 3 (line 4)
  7. fp -> frame 2 (line 5)
  8. test on line 8 is TRUE
  9. fp -> frame 1 (line 14)
  10. line 18 loop back up
  11. fpprev -> frame 1 (line 4)
  12. fp -> NULL (line 5)
  13. test on line 8 is FALSE
  14. loop exit due to fp == NULL (line 18)

At loop exit, fp -> NULL and fpprev -> frame 1 (the M base frame).

The mv_stent loop, since fp is NULL, would use fpprev (the call-in base frame) in its end-point test.



Now - new code:

1        do
2        {
3               ... <irrelevant to this discussion stuff>
4               fpprev = fp;
5               if (done)
6                       break;
7               if (SFT_CI & fp->flags)
8               {       /* Callins needs to be able to crawl past apparent end of stack to earlier stack segments.
9                        * We should be in the base frame now. See if an earlier frame exists.
10                       * Note we don't worry about trigger base frames here because triggers *always* have a
11                       * different symbol table - previous symbol tables and stack levels are not affected.
12                       */
13                      fp = *(stack_frame **)(fp + 1); /* Backups up to "prev pointer" created by base_frame() */
14                      if ((NULL == fp) || (fp >= (stack_frame *)stackbase) || (fp < (stack_frame *)stacktop))
15                              break;  /* Pointer not within the stack -- must be earliest occurence */
16              } else
17                      fp = fp->old_frame_pointer;
18      } while(fp);
19      /* Next, check the mv_stents for the stackframes we processed. Certain mv_stents also have hash
20       * table references in them that need repair.
21       */
22      for (mv_st_ent = mv_chain;
23           mv_st_ent < (mv_stent *)(fp ? fp : fpprev);        /* Last stack frame actually processed */
24           mv_st_ent = (mv_stent *)(mv_st_ent->mv_st_next + (char *)mv_st_ent))
25      {
26              ... <rest of loop>
27      }

Recall that the new stack looks like:
[1:M-base-frame (no prev frame)] : [2:callin-base-frame (has SFT_CI set)] : [3:callin-routine-frame]

Operations in loop:
  1. fp -> frame 1 (line 1)
  2. fpprev -> frame 1 (line 4)
  3. test on line 7 is FALSE
  4. fp -> frame 2 (line 17)
  5. line 18 loop back up
  6. fpprev -> frame 2 (line 4)
  7. test on line 7 is TRUE
  8. fp -> frame 1 (line 13)
  9. line 18 loop back up
  10. fpprev -> frame 1 (line 4)
  11. test on line 7 is FALSE
  12. fp -> NULL (line 17)
  13. loop exit due to fp == NULL (line 18)

At loop exit, fp -> NULL and fpprev -> frame 1 (the M base frame).

The mv_stent loop, since fp is NULL would use fpprev (the call-in base frame) in its end-point test.

******************************************************************************************************

This is if a call-in frame is NOT present in the stack.

Using the previous code above and the following M stack:

[1:M-base-frame (no prev frame)] : [2:M-frameA] : [3:M-frameB]

Operations in loop:
  1. fp -> frame 3 (line 1)
  2. fpprev -> frame 3 (line 4)
  3. fp -> frame 2 (line 5)
  4. test on line 8 is FALSE
  5. line 18 loop back up
  6. fpprev = frame 2 (line 4)
  7. fp -> frame 1 (line 5)
  8. test on line 8 is FALSE
  9. fpprev -> frame 1 (line 4)
  10. fp -> NULL (line 5)
  11. test on line 8 is FALSE
  12. loop exit due to fp == NULL (line 18)

At loop exit, fp -> NULL and fpprev -> frame 1 (the M base frame).

The mv_stent loop, since fp is NULL would use fpprev (the M base frame) in its end-point test.


Using the newer code above and the same M stack:

Operations in loop:
  1. fp -> frame 3 (line 1)
  2. fpprev -> frame 3 (line 4)
  3. test on line 7 is FALSE
  4. fp -> frame 2 (line 17)
  5. line 18 loop back up
  6. fpprev -> frame 2 (line 4)
  7. test on line 7 is FALSE
  8. fp -> frame 1 (line 17)
  9. line 18 loop back up
  10. fpprev -> frame 1 (line 4)
  11. test on line 7 is FALSE
  12. fp -> NULL
  13. lop exit due to fp == NULL (line 18)

At loop exit, fp -> NULL and fpprev -> frame 1 (the M base frame).

The mv_stent loop, since fp is NULL would use fpprev (the M base frame) in its end-point test.

******************************************************************************************************

I'm not seeing the problem you are theorizing. Can you elaborate?
```

Reply:
After our phone conversation, I went back and see what you mean about when "done" is set, we could end on a different stack frame than previously. Figuring out whether that's bad or not will take some significant time so instead I have put the loop back the way it was and made adjustments for the lack of GTM$CI frame a different way that won't impact the values at loop exit. Were that actually the source of a bug (which seems likely but no idea how to trip it at this time), it would have been hard to find..